optbuilder: Fix bug with column visibility

rytaft opened the pull request:
Before this fix, invalid queries such as
```
  SELECT c FROM (SELECT a FROM abc)
```
were (incorrectly) not causing an error in the optbuilder.
This commit fixes the problem by ensuring that FROM columns are
not visible outside of the subquery unless they are included
in the SELECT list.

There is an additional complexity for subqueries
with an order by clause such as
```
  SELECT c FROM (SELECT a FROM abc ORDER BY c)
```
Since column c is included in the subquery projection, it must be
hidden and removed from the scope before the names in the outer
query are resolved.

Release note: None

rytaft committed:
optbuilder: Fix bug with column visibility

Before this fix, invalid queries such as
  SELECT c FROM (SELECT a FROM abc)
were (incorrectly) not causing an error in the optbuilder.
This commit fixes the problem by ensuring that FROM columns are
not visible outside of the subquery unless they are included
in the SELECT list.

There is an additional complexity for subqueries
with an order by clause such as
  SELECT c FROM (SELECT a FROM abc ORDER BY c)
Since column c is included in the subquery projection, it must be
hidden and removed from the scope before the names in the outer
query are resolved.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24091)
<!-- Reviewable:end -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/testdata/aggregate, line 1771 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84arxyCISUDLmOKg_y:-L84arxyCISUDLmOKg_z:bodxwxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad2edbb955a8ece78aa9d248da0649ac728132e/pkg/sql/opt/optbuilder/testdata/aggregate#L1771)):*
> ```
>  ├── columns: column5:5(string)
>  ├── project
>  │    ├── columns: kv.s:4(string)
> ```

If we run with normalization rules, does this get added back? (I assume yes, or there'd probably be changes in the execbuilder tests).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-:-L84b9Ib4NJpjbZuyG2W:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/select.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84cfeXeBPAKAfSmLUl:-L84cfeYJk_KIeEdn9A_:bsan1u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/940b7c1011ca549ac8739c06a1fb0617d099cffb/pkg/sql/opt/optbuilder/select.go#L192)):*
> ```Go
> 	}
> 
> 	// Remove hidden columns from outScope.
> ```

I assume this works correctly with `rowid` cols? Are there tests which verify that?

---

*[pkg/sql/opt/optbuilder/select.go, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84cuiZGzihol7p4SKJ:-L84cuiZGzihol7p4SKK:b-a43qq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/940b7c1011ca549ac8739c06a1fb0617d099cffb/pkg/sql/opt/optbuilder/select.go#L194)):*
> ```Go
> 	// Remove hidden columns from outScope.
> 	n := 0
> 	for i := range outScope.cols {
> ```

This feels like it should be a `removeHiddenCols` method on the `scope` class, perhaps?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-:-L84d2vknHudUSGf996L:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
TFTRs! Fixed a few more bugs - ready for another look.

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/opt/exec/execbuilder/testdata/orderby, line 917 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84n-eummmeepeTNJUS:-L84n-eummmeepeTNJUT:b-b5rv3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/99360d003c9b304927a1f81bdeccb426a638aa47/pkg/sql/opt/exec/execbuilder/testdata/orderby#L917)):*
> ```
> 
> exec-explain
> SELECT * FROM (SELECT * FROM (VALUES ('a'), ('b'), ('c')) AS c(x) ORDER BY x) WITH ORDINALITY
> ```

We haven't implemented `WITH ORDINALITY` yet -- the sort will get added back at that point.

---

*[pkg/sql/opt/optbuilder/select.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84cfeXeBPAKAfSmLUl:-L84fgEQbfe_qb_QPgoY:b-4qg8vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/940b7c1011ca549ac8739c06a1fb0617d099cffb/pkg/sql/opt/optbuilder/select.go#L192)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I assume this works correctly with `rowid` cols? Are there tests which verify that?
</blockquote></details>

Good catch - this code was in the wrong place. Now it's up on line 66 (inside the `tree.Subquery` case of `buildTable`). Added a test as well.

---

*[pkg/sql/opt/optbuilder/select.go, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84cuiZGzihol7p4SKJ:-L84ffsQwz3AsY359jVj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/940b7c1011ca549ac8739c06a1fb0617d099cffb/pkg/sql/opt/optbuilder/select.go#L194)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This feels like it should be a `removeHiddenCols` method on the `scope` class, perhaps?
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/testdata/aggregate, line 1771 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84arxyCISUDLmOKg_y:-L84fqEWxUN1KK2OR8-F:b-hyw9oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad2edbb955a8ece78aa9d248da0649ac728132e/pkg/sql/opt/optbuilder/testdata/aggregate#L1771)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

If we run with normalization rules, does this get added back? (I assume yes, or there'd probably be changes in the execbuilder tests).
</blockquote></details>

There were actually changes in the execbuilder tests (I forgot to update them), but not for these queries.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-:-L84oebfuP_d6N-3EpLy:bsadyqt)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/exec/execbuilder/testdata/orderby, line 880 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L85HFA4YvBWl6PNYCtu:-L85HFA4YvBWl6PNYCtv:b-p7ed39) ([raw file](https://github.com/cockroachdb/cockroach/blob/99360d003c9b304927a1f81bdeccb426a638aa47/pkg/sql/opt/exec/execbuilder/testdata/orderby#L880)):*
> ```
> 
> exec-explain
> SELECT * FROM (SELECT * FROM (VALUES ('a'), ('b'), ('c')) AS c(x) ORDER BY x)
> ```

We are ignoring the "inner" ORDER BY.  This is probably fine (even though CRDB currently doesn't ignore it, nor does PG). Can you add another testcase above without the outer `SELECT *` to make sure we do sort the values?

---

*[pkg/sql/opt/exec/execbuilder/testdata/orderby, line 917 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091#-L84n-eummmeepeTNJUS:-L85Hz4ZwPKf9ywtvwFs:b-z3t6es) ([raw file](https://github.com/cockroachdb/cockroach/blob/99360d003c9b304927a1f81bdeccb426a638aa47/pkg/sql/opt/exec/execbuilder/testdata/orderby#L917)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

We haven't implemented `WITH ORDINALITY` yet -- the sort will get added back at that point.
</blockquote></details>

I would comment it out until then and mention that once supported this should sort.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24091)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/opt/exec/execbuilder/testdata/orderby, line 880 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24091#-L85HFA4YvBWl6PNYCtu:-L87jG-8B32tmkkn6DvO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99360d003c9b304927a1f81bdeccb426a638aa47/pkg/sql/opt/exec/execbuilder/testdata/orderby#L880)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We are ignoring the "inner" ORDER BY.  This is probably fine (even though CRDB currently doesn't ignore it, nor does PG). Can you add another testcase above without the outer `SELECT *` to make sure we do sort the values?
</blockquote></details>

Done.

---

*[pkg/sql/opt/exec/execbuilder/testdata/orderby, line 917 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24091#-L84n-eummmeepeTNJUS:-L87jLapEoJfEyx4237r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99360d003c9b304927a1f81bdeccb426a638aa47/pkg/sql/opt/exec/execbuilder/testdata/orderby#L917)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would comment it out until then and mention that once supported this should sort.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24091)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewed 1 of 4 files at r1, 4 of 6 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24091)*
<!-- Sent from Reviewable.io -->
